core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+    }
+
+    if (timeLapse.isPresent() && timeLapse.get() instanceof StateFragment
+        && ((StateFragment) timeLapse.get()).getFinish() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Compute EventStructure for ExecutionEnd.
+   * 
+   * @param ancestors
+   *          Deque<EObject>
+   * @param result
+   *          List<EventContext>
+   * @param end
+   *          ExecutionEnd
+   */
+  private static void visit(Deque<CapellaElement> ancestors, List<EventContext> result, ExecutionEnd end,
+      Optional<TimeLapse> timeLapse) {
+    // Should not happen in current implementation of Capella Sequence diagrams
+    if (timeLapse.isPresent() && timeLapse.get().getStart() == end) {
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), true, ancestors.size() + 1));
+      ancestors.push(timeLapse.get());
+    }
+
+    // Execution without return branch end / ASyncCall
+    if (timeLapse.isPresent() && timeLapse.get().getFinish() == end) {
+      ancestors.pop();
+      result.add(new EventContext(ancestors.peek(), timeLapse.get(), false, ancestors.size() + 1));
+    }
+  }
+
+  /**
+   * Return covered instance role.
+   * 
+   * @param end
+   *          InteractionFragment
+   * @return covered instance role.
+   */
+  private static InstanceRole getCoveredInstanceRole(InteractionFragment end) {
+    InstanceRole covered = null;
+    if (end instanceof AbstractEnd) {
+      covered = getInteractionCache(AbstractEnd::getCovered, (AbstractEnd) end);
+    } else if (end instanceof InteractionState) {
+      covered = getInteractionCache(InteractionState::getCovered, (InteractionState) end);
+    }
+    return covered;
+  }
+
+  /**
+   * Compute cache for TimeLapse start and finish ends.
+   * 
+   * @param scenario
+   *          Scenario
+   */
+  private static void computeTimeLapseStartAndEndCache(Scenario scenario) {
+    // cache Execution.start -> Execution and Execution.finish -> Execution
+    // scan timelapse only one time
+    if (getInteractionFragmentToTimeLapseCache().isEmpty()) {
+      Stream<TimeLapse> timeLapses = scenario.getOwnedTimeLapses().stream()
+          .filter(tl -> !(tl instanceof AbstractFragment));
+      timeLapses.forEach(e -> {
+        // Keep Interaction to TimeLapse info in cache
+        // M2 allows to retrieve an Execution from an ExecutionEnd (execution.getFinish() when there is no
+        // return branch : ASyncCall)
+        // but not an Execution from a MessagEnd (SyncCall and ASyncCall) or a StateFragment from an
+        // InteractionState.
+        putTimeLapseInCache(e.getStart(), e);
+        putTimeLapseInCache(e.getFinish(), e);
+      });
+    }
+  }
+
+  /**
+   * Helper class to keep track of who "contains" who depending on the interleaving of the start/finish ends.
+   * 
+   * @author pcdavid
+   */
+  public static final class EventContext {
```

Comment:
The name of this class is confusing. There is already an `Event` class in Capella that has different meaning.

A more suitable name  would be `Sibling` because this class contains siblings of InstanceRoles (closely related elements).
An sibling contains a `parent` which can be used to filter the `children`.

Reply:
Hi Sandu, 
We kept Event,  EventContext and EventEnd as this is has a meaning in the Sirius Sequence diagram (used in layout, internal api ISequenceEvent), it was really helpfull in debug to have the same wording than in Sirius (refresh, layout) and not new vocabulary. 
It was also a way to be sure to avoid to reuse TimeLapse, Fragments, ...

Reply:
Hi Sandu. 

After discussion with Philippe, it is ok to rename the EventContext into something else, to avoid confusion with the "event" from the Capella M2 to help the maintainers of this code, but "sibling" does not look appropriate: 

> definition of sibling: each of two or more children or offspring having one or both parents in common; a brother or sister.

IMO, we should try to find another name, as the EventContext elements are not just siblings under the same parent, but their organization represents the hierarchical tree of Executions/States on the diagram and the structure is used from the VSM to build/find the semantic candidates for level/depth in evaluation of the mappings's expression during the SequenceDDiagram refresh.